ui/app/selectors/confirm-transaction.js

```diff
@@ -63,3 +64,101 @@ export const unconfirmedTransactionsHashSelector = createSelector(
 
 export const currentCurrencySelector = state => state.metamask.currentCurrency
 export const conversionRateSelector = state => state.metamask.conversionRate
+
+const txDataSelector = state => state.confirmTransaction.txData
+const tokenDataSelector = state => state.confirmTransaction.tokenData
+const tokenPropsSelector = state => state.confirmTransaction.tokenProps
+
+const contractExchangeRatesSelector = state => state.metamask.contractExchangeRates
+
+const tokenDecimalsSelector = createSelector(
+  tokenPropsSelector,
+  tokenProps => tokenProps && tokenProps.tokenDecimals
+)
+
+const tokenDataParamsSelector = createSelector(
+  tokenDataSelector,
+  tokenData => tokenData && tokenData.params || []
+)
+
+const txParamsSelector = createSelector(
+  txDataSelector,
+  txData => txData && txData.txParams || {}
+)
+
+export const tokenAddressSelector = createSelector(
+  txParamsSelector,
+  txParams => txParams && txParams.to
+)
+
+const TOKEN_PARAM_SPENDER = '_spender'
+const TOKEN_PARAM_TO = '_to'
+const TOKEN_PARAM_VALUE = '_value'
+
+export const tokenAmountAndToAddressSelector = createSelector(
+  tokenDataParamsSelector,
+  params => {
+    let toAddress = ''
+    let tokenAmount = 0
+
+    if (params && params.length) {
+      const toParam = params.find(param => param.name === TOKEN_PARAM_TO)
+      const valueParam = params.find(param => param.name === TOKEN_PARAM_VALUE)
+      toAddress = toParam ? toParam.value : params[0].value
+      tokenAmount = valueParam ? +valueParam.value : +params[1].value
+    }
+
+    return {
+      toAddress,
+      tokenAmount,
+    }
+  }
+)
+
+export const approveTokenAmountAndToAddressSelector = createSelector(
+  tokenDataParamsSelector,
+  params => {
+    let toAddress = ''
+    let tokenAmount = 0
+
+    if (params && params.length) {
+      toAddress = params.find(param => param.name === TOKEN_PARAM_SPENDER).value
+      tokenAmount = +params.find(param => param.name === TOKEN_PARAM_VALUE).value
```

Comment:
I would recommend using the Number constructor here instead of the implicit conversion